File: frontend/src/types/booking.ts
/**
 * Type definitions matching backend Pydantic schemas
 * These are strongly typed to ensure type safety across the application
 */

export type BookingStatus =
  | 'PENDING'
  | 'CONFIRMED'
  | 'IN_PROGRESS'
  | 'COMPLETED'
  | 'CANCELLED';

export interface BookingBase {
  customer_name: string;
  customer_email: string;
  customer_phone: string;
  move_date: string; // ISO 8601 datetime
  pickup_address: string;
  pickup_city: string;
  pickup_state: string;
  pickup_zip: string;
  dropoff_address: string;
  dropoff_city: string;
  dropoff_state: string;
  dropoff_zip: string;
  estimated_distance_miles: number;
  estimated_duration_hours: number;
  special_items: string[];
  pickup_floors: number;
  dropoff_floors: number;
  has_elevator_pickup: boolean;
  has_elevator_dropoff: boolean;
  customer_notes?: string;
}

export interface BookingCreate extends BookingBase {}

export interface BookingResponse extends BookingBase {
  id: string; // UUID
  org_id: string; // UUID
  truck_id: string | null; // UUID
  status: BookingStatus;
  effective_start: string; // ISO 8601 datetime
  effective_end: string; // ISO 8601 datetime
  created_at: string; // ISO 8601 datetime
  updated_at: string; // ISO 8601 datetime;
}

export interface AvailabilityCheck {
  org_id: string; // UUID
  truck_id: string | null; // UUID
  move_date: string; // ISO 8601 datetime
  estimated_duration_hours: number;
}

export interface AvailabilityResponse {
  available: boolean;
  conflicting_bookings: BookingResponse[];
}

export interface PriceEstimate {
  estimated_amount: number;
  platform_fee: number;
  breakdown: PriceBreakdown;
}

export interface PriceBreakdown {
  base_hourly_cost: number;
  base_mileage_cost: number;
  surcharges: Surcharge[];
  subtotal: number;
  minimum_applied: boolean;
  total: number;
}

export interface Surcharge {
  type: string;
  applied: boolean;
  amount: number;
  description?: string;
  flights?: number;
  multiplier?: number;
}

export interface TruckDetails {
  id: string;
  license_plate: string;
  make: string;
  model: string;
  year: number;
  capacity_cubic_feet: number;
}

export interface DriverDetails {
  id: string;
  first_name: string;
  last_name: string;
  phone: string;
  photo_url: string | null;
}

export interface BookingWithDetails extends BookingResponse {
  truck?: TruckDetails;
  driver?: DriverDetails;
  price_estimate?: PriceEstimate;
}

File: app/schemas/booking.py
"""Booking schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import EmailStr, Field, field_validator

from app.models.booking import BookingStatus
from app.schemas.base import BaseSchema, ResourceResponse


class BookingBase(BaseSchema):
    """Base booking schema."""

    customer_name: str = Field(..., min_length=1, max_length=255)
    customer_email: EmailStr
    customer_phone: str = Field(..., pattern=r"^\+?1?\d{9,15}$")
    move_date: datetime = Field(..., description="Requested move date and time")
    pickup_address: str = Field(..., min_length=1, max_length=512)
    pickup_city: str = Field(..., min_length=1, max_length=100)
    pickup_state: str = Field(..., min_length=2, max_length=50)
    pickup_zip: str = Field(..., pattern=r"^\d{5}(-\d{4})?$")
    dropoff_address: str = Field(..., min_length=1, max_length=512)
    dropoff_city: str = Field(..., min_length=1, max_length=100)
    dropoff_state: str = Field(..., min_length=2, max_length=50)
    dropoff_zip: str = Field(..., pattern=r"^\d{5}(-\d{4})?$")
    estimated_distance_miles: float = Field(..., gt=0)
    estimated_duration_hours: float = Field(..., gt=0, le=24)
    special_items: list[str] = Field(default_factory=list)
    pickup_floors: int = Field(default=0, ge=0, le=100)
    dropoff_floors: int = Field(default=0, ge=0, le=100)
    has_elevator_pickup: bool = False
    has_elevator_dropoff: bool = False
    customer_notes: str | None = Field(None, max_length=2000)

    @field_validator("move_date")
    @classmethod
    def validate_move_date(cls, v: datetime) -> datetime:
        """Validate move date is in the future."""
        if v < datetime.now():
            raise ValueError("Move date must be in the future")
        return v


class BookingCreate(BookingBase):
    """Schema for creating a booking."""

    truck_id: UUID
    org_id: UUID


class BookingUpdate(BaseSchema):
    """Schema for updating a booking."""

    move_date: datetime | None = None
    status: BookingStatus | None = None
    final_amount: float | None = Field(None, ge=0)
    internal_notes: str | None = Field(None, max_length=2000)


class BookingResponse(BookingBase, ResourceResponse):
    """Schema for booking response."""

    id: UUID
    org_id: UUID
    truck_id: UUID
    commute_buffer_minutes: int
    effective_start: datetime
    effective_end: datetime
    estimated_amount: float
    final_amount: float | None = None
    platform_fee: float
    stripe_payment_intent_id: str | None = None
    status: BookingStatus
    internal_notes: str | None = None


class AvailabilityCheck(BaseSchema):
    """Schema for checking truck availability."""

    truck_id: UUID
    move_date: datetime
    estimated_duration_hours: float = Field(..., gt=0, le=24)
    commute_buffer_minutes: int = Field(default=30, ge=0, le=120)


class AvailabilitySlot(BaseSchema):
    """Available time slot."""

    start: datetime
    end: datetime


class AvailabilityResponse(BaseSchema):
    """Schema for availability response."""

    truck_id: UUID
    is_available: bool
    requested_start: datetime
    requested_end: datetime
    conflicts: list[BookingResponse] = Field(default_factory=list)
    suggested_slots: list[AvailabilitySlot] = Field(default_factory=list)
